pandas/_libs/src/ujson/python/JSONtoObj.c

```diff
@@ -119,6 +122,9 @@ JSOBJ Object_npyNewArray(void *prv, void *_decoder) {
         npyarr->labels[0] = npyarr->labels[1] = NULL;
 
         npyarr->shape.ptr = PyObject_Malloc(sizeof(npy_intp) * NPY_MAXDIMS);
+        if (npyarr->shape.ptr == NULL) {
```

Comment:
At this point in the code you have malloc'ed decoder->npyarr, so you also need to free it (using PyObject_Free instead of the built-in)